### PR TITLE
ASR models knowledge distillation example

### PR DESCRIPTION
**Pull request by DavitGrigoryan132**
# What does this PR do ?

In this PR I am adding an example of script for knowledge distillation of EncDecCTCModels which can be adapted for all ASR models, and an example of config file to run knowledge distillation.

**Collection**: ASR

# Changelog

- Created a knowledge distillation module for EncDecCTCModels
- Created a config file scheme for comfortable knowledge distillation, where you can specify between which layers you want to put the loss function and also add an adapter layer between them if teacher and student models features dimensions are not matching on that layers

# Usage

You can simply run the example using this command
```python
python examples/asr/distillation/distill.py --config-path=. --config-name=distill.yaml
```

You need to adapt config file for your models and datasets before running the script
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

### CODE REVIEW

**Comment by DavitGrigoryan132** [comment]
Hey @titu1994 can you please review the changes?

File: examples/asr/distillation/distill.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+    def forward(self, t_feats, s_feats):
+        x = t_feats.permute(0, 2, 1)
+        x = self.adapter(x)
+        x = torch.nn.functional.interpolate(x, size=s_feats.size(1), mode='linear', align_corners=False)
+        proj_t = x.permute(0, 2, 1)
+        return self.kd_loss_fn(proj_t, s_feats)
+
+
+class KnowledgeDistillationModule(EncDecCTCModel):
```
Comment (titu1994) [inline examples/asr/distillation/distill.py]: Could you rename this KnowledgeDistillationCTCModule?

**Comment by titu1994** [comment]
Sorry for the delay. Overall logic seems to be correct. Minor nit on the naming, but we can merge it and rename it later on

**Review by titu1994 — APPROVED** [review]

**Comment by DavitGrigoryan132** [comment]
@titu1994 Thank you very much, I renamed it as you said

**Comment by DavitGrigoryan132** [comment]
@titu1994 @ko3n1g Hi, can we please merge the PR as it got approved? I renamed the class after your comment and now everything should be correct
